### PR TITLE
update batch_mode to handle non constant matrices

### PR DESCRIPTION
**Pull request by bcordo**
This PR:
1. Adds non-constant F,H, and Q matrices to the batch filter
2. Adds non-constant F matrix to the rts smoother

Only for linear filters. 

@JohnEmhoff or Sanujit  could you hook us up with a review.

### CODE REVIEW

File: filterpy/kalman/kalman_filter.py [hunk 1]
```diff
@@ -304,49 +304,57 @@ def predict(self, u=0):
...
+            a value of None in any position will cause the filter
+            to use `self.R` for that time step.
+        Hs : list-like, optional
+            optional list of values to use for the measurement matrix;
+            a value of None in any position will cause the filter
+            to use `self.R` for that time step.
+        Qs : list-like, optional
+            optional list of values to use for the process error
+            covariance; a value of None in any position will cause the filter
+            to use `self.R` for that time step.
```
Comment (relaxbox) [inline filterpy/kalman/kalman_filter.py:355]: I think most of these comments intend to reference the respective parameter instead `self.R`

File: filterpy/kalman/kalman_filter.py [hunk 2]
```diff
@@ -304,49 +304,57 @@ def predict(self, u=0):
...
-
         covariance_predictions: np.array((n,dim_x,dim_x))
             array of the covariances for each time step after the prediction.
             In other words `covariance[k,:,:]` is the covariance at step `k`.
         """
 
         n = np.size(zs,0)
-        if Rs is None:
-            Rs = [None]*n
+        if np.all(Rs == None):
```
Comment (relaxbox) [inline filterpy/kalman/kalman_filter.py]: These checks look a little strange to me, I think something like `if Rs is None` is the more canonical way to check for None-ness

**Comment by relaxbox** [comment]
Looks good to me, seems like you were able to fit everything in following the style of the original author. I bet he'd appreciate a pull request.

**Comment by bcordo** [comment]
I realized when looking back over the code that, the new implementation of batch_filter has a side effect of altering self.F and self.Q. This, along with exceptions, can lead to the filter ending up in some unknown state.

Instead `predict()` should accept an optional F and Q, since both matrices are allowed to be time varying. This would have the benefit of being more flexible and simplifying the logic of batch_filter.

I also added a dynamic B, and u to handle the general case, along with some formatting changes.

**Comment by relaxbox** [comment]
Looks good to me -- I say merge it up!

**Pull request merged**